staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go

```diff
@@ -580,55 +678,70 @@ func mergeMap(original, patch map[string]interface{}, t reflect.Type, mergeDelet
-			if err != nil {
-				return nil, err
-			}
-
-			if originalType.Kind() == reflect.Map && fieldPatchStrategy != replaceDirective {
-				typedOriginal := original[k].(map[string]interface{})
-				typedPatch := patchV.(map[string]interface{})
-				var err error
-				original[k], err = mergeMap(typedOriginal, typedPatch, fieldType, mergeDeleteList, ignoreUnmatchedNulls)
-				if err != nil {
-					return nil, err
-				}
-
-				continue
-			}
+		if originalType != patchType {
+			original[k] = patchV
+			continue
+		}
+		// If they're both maps or lists, recurse into the value.
+		// First find the fieldPatchStrategy and fieldPatchMergeKey.
+		fieldType, fieldPatchStrategy, fieldPatchMergeKey, err := forkedjson.LookupPatchMetadata(t, k)
+		if err != nil {
+			return nil, err
+		}
+		switch originalType.Kind() {
+		case reflect.Map:
+			err = mergeMapHandler(original, k, patchV, fieldType, fieldPatchStrategy, mergeOptions)
+		case reflect.Slice:
+			err = mergeSliceHandler(original, k, patchV, fieldType, fieldPatchStrategy, fieldPatchMergeKey, isDeleteList, mergeOptions)
+		default:
+			original[k] = patchV
+		}
 
-			if originalType.Kind() == reflect.Slice && fieldPatchStrategy == mergeDirective {
-				elemType := fieldType.Elem()
-				typedOriginal := original[k].([]interface{})
-				typedPatch := patchV.([]interface{})
-				var err error
-				original[k], err = mergeSlice(typedOriginal, typedPatch, elemType, fieldPatchMergeKey, mergeDeleteList, isDeleteList, ignoreUnmatchedNulls)
-				if err != nil {
-					return nil, err
-				}
+	}
+	return original, nil
+}
 
-				continue
-			}
+// mergeMapHandler handles how to merge `patchV` whose key is `key` with `original` respecting
+// fieldPatchStrategy and mergeOptions.
+func mergeMapHandler(original map[string]interface{}, key string, patchV interface{}, fieldType reflect.Type,
+	fieldPatchStrategy string, mergeOptions MergeOptions) error {
+	if fieldPatchStrategy != replaceDirective {
+		typedOriginal := original[key].(map[string]interface{})
+		typedPatch := patchV.(map[string]interface{})
+		var err error
+		if original[key], err = mergeMap(typedOriginal, typedPatch, fieldType, mergeOptions); err != nil {
+			return err
 		}
-
-		// If originalType and patchType are different OR the types are both
-		// maps or slices but we're just supposed to replace them, just take
-		// the value from patch.
-		original[k] = patchV
+	} else {
+		original[key] = patchV
 	}
+	return nil
+}
 
-	return original, nil
+// mergeMapHandler handles how to merge `patchV` whose key is `key` with `original` respecting
+// fieldPatchStrategy, fieldPatchMergeKey, isDeleteList and mergeOptions.
+func mergeSliceHandler(original map[string]interface{}, key string, patchV interface{}, fieldType reflect.Type,
+	fieldPatchStrategy, fieldPatchMergeKey string, isDeleteList bool, mergeOptions MergeOptions) error {
+	if fieldPatchStrategy == mergeDirective {
+		elemType := fieldType.Elem()
```

Comment:
I want to keep the signature of `mergeSliceHandler` and `mergeSliceHandler` similar. So I prefer to use  `fieldType`